### ISSUE TITLE
Better form abstraction

### ISSUE THREAD

**Issue by gsidebo** [mitodl/open-discussions#1570]
PR https://github.com/mitodl/open-discussions/pull/1485 was a good indicator that we could use a better abstraction for including forms in our components. It required quite a lot of boilerplate to develop and test. The [`withForm` HOC](https://github.com/mitodl/open-discussions/blob/v0.56.1/static/js/hoc/withForm.js) has been useful but it has at least 2 drawbacks: (1) you can't use it to add more than one form in a component, and (2) it relies on several props that you need to configure in the wrapped component, which results in us splitting a significant amount of form behavior logic between the form component and the page component that includes it.

**Comment by aliceriot** [comment]
I agree, I think in general as we add more HOCs to pages things get a lot more difficult to understand and debug. some of our container components now have as many as six HOCs wrapping them, which is just far too many I think.

It would be good to look at third-party libraries and see if there's anything well-built and well-maintained (maybe designed to work with react+redux) that we could depend on instead.

**Comment by gsidebo** [comment]
Personally I don't have a problem (yet) with a high # of HOCs as long as the interdependence between the wrapped component and wrapping component is kept to a minimum. To me it's good a composition-over-inheritance pattern to use. I haven't yet had a lot of debugging trouble due to HOCs but if we run into some, it would be great as a case study. It's a nice pattern but I wouldn't be surprised to find some major drawbacks if we overuse/misuse it

**Comment by aliceriot** [comment]
there's a few libraries out there, here's a few I looked at when looking around quickly:

- [redux-form](https://redux-form.com/): this one seems to be a pretty solid react+redux choice. it's based on a HOC API, and I'm not sure what it gives us that we don't really already have. not sure if it supports multiple forms on the same page
- [formik](https://github.com/jaredpalmer/formik): this one seems interesting, it's a component-based API instead of a HOC, which means that I think it would be easy to put multiple forms on one page.
